File: app/src/main/java/samples/despotoski/nikola/com/bottomsheetsample/BottomSheetActivity.java
package samples.despotoski.nikola.com.bottomsheetsample;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import java.util.ArrayList;
import java.util.List;

public class BottomSheetActivity extends AppCompatActivity {

    private TextView mStateText;
    private TextView mOffsetText;
    private BottomSheetBehavior mBottomSheetBehavior;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_bottom_sheet);
        final RecyclerView recyclerView = (RecyclerView) findViewById(R.id.recyclerview);
        mOffsetText = (TextView) findViewById(R.id.offsetText);
        mStateText = (TextView) findViewById(R.id.stateText);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        final LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setAutoMeasureEnabled(true);
        ApplicationAdapter adapter = new ApplicationAdapter(this, listApplications(this));
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setAdapter(adapter);
        FrameLayout parentThatHasBottomSheetBehavior = (FrameLayout) recyclerView.getParent().getParent();
        mBottomSheetBehavior = BottomSheetBehavior.from(parentThatHasBottomSheetBehavior);
        if (mBottomSheetBehavior != null) {
            setStateText(mBottomSheetBehavior.getState());
            mBottomSheetBehavior.setBottomSheetCallback(new BottomSheetBehavior.BottomSheetCallback() {
                @Override
                public void onStateChanged(@NonNull View bottomSheet, int newState) {
                    setStateText(newState);
                }

                @Override
                public void onSlide(@NonNull View bottomSheet, float slideOffset) {
                    setOffsetText(slideOffset);
                }
            });
        }

        View peekButton = findViewById(R.id.peek_me);
        peekButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //Let's peek it, programmatically
                View peakView = findViewById(R.id.drag_me);
                mBottomSheetBehavior.setPeekHeight(peakView.getHeight());
                peakView.requestLayout();
            }
        });
        View modal = findViewById(R.id.as_modal);
        modal.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                BottomSheetDialogFragment bottomSheetDialogFragment = new CustomBottomSheetDialogFragment();
                bottomSheetDialogFragment.show(getSupportFragmentManager(), bottomSheetDialogFragment.getTag());
            }
        });
    }

    @Override
    public void onBackPressed() {
        if (mBottomSheetBehavior.getState() != BottomSheetBehavior.STATE_HIDDEN) {
            mBottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
        } else {
            super.onBackPressed();
        }
    }

    private void setOffsetText(float slideOffset) {
        mOffsetText.setText(getString(R.string.offset, slideOffset));
    }

    private void setStateText(int newState) {
        mStateText.setText(getStateAsString(newState));
    }

    public static int getStateAsString(int newState) {
        switch (newState) {
            case BottomSheetBehavior.STATE_COLLAPSED:
                return R.string.collapsed;
            case BottomSheetBehavior.STATE_DRAGGING:
                return R.string.dragging;
            case BottomSheetBehavior.STATE_EXPANDED:
                return R.string.expanded;
            case BottomSheetBehavior.STATE_HIDDEN:
                return R.string.hidden;
            case BottomSheetBehavior.STATE_SETTLING:
                return R.string.settling;
        }
        return R.string.undefined;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_bottom_sheet, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }


    public static List<ApplicationInfo> listApplications(Context context) {
        int flags = PackageManager.GET_META_DATA;
        List<ApplicationInfo> installedApps = new ArrayList<>();
        PackageManager pm = context.getPackageManager();
        List<ApplicationInfo> applications = pm.getInstalledApplications(flags);
        for (ApplicationInfo appInfo : applications) {
            if ((appInfo.flags & ApplicationInfo.FLAG_SYSTEM) != 1) {
                installedApps.add(appInfo);
            }
        }
        return installedApps;
    }
}

File: app/src/main/java/samples/despotoski/nikola/com/bottomsheetsample/BottomSheetActivity.kt
package samples.despotoski.nikola.com.bottomsheetsample

import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.FrameLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetBehavior.BottomSheetCallback
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import kotlinx.android.synthetic.main.bottom_sheet_1.view.*
import samples.despotoski.nikola.com.bottomsheetsample.databinding.ActivityBottomSheetBinding
import java.util.*

class BottomSheetActivity : AppCompatActivity() {
    private var mBottomSheetBehavior: BottomSheetBehavior<*>? = null

    lateinit var binding : ActivityBottomSheetBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = DataBindingUtil.setContentView(this,R.layout.activity_bottom_sheet)

        setSupportActionBar(binding.toolbar)

        setupBottomSheet1()

        setupBottomSheet2()
    }

    private fun setupBottomSheet2() {
        binding.showBottomSheet2.setOnClickListener {
            val bottomSheetDialogFragment: BottomSheetDialogFragment = CustomBottomSheetDialogFragment()
            bottomSheetDialogFragment.show(supportFragmentManager, bottomSheetDialogFragment.tag)
        }
    }

    private fun setupBottomSheet1() {
        binding.showBottomSheet1.setOnClickListener { //Let's peek it, programmatically
            mBottomSheetBehavior?.peekHeight = binding.layoutBottomSheet1.textViewDragMe.height
            binding.layoutBottomSheet1.textViewDragMe.requestLayout()
        }

        val recyclerView = binding.layoutBottomSheet1.recyclerview
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = ApplicationAdapter(this, listApplications(this))
        val parentThatHasBottomSheetBehavior = recyclerView.parent.parent as FrameLayout
        mBottomSheetBehavior = BottomSheetBehavior.from(parentThatHasBottomSheetBehavior)
        mBottomSheetBehavior?.let {
            binding.stateText.setText(getStateAsString(it.state))
            it.setBottomSheetCallback(object : BottomSheetCallback() {
                override fun onStateChanged(bottomSheet: View, newState: Int) {
                    binding.stateText.setText(getStateAsString(newState))
                }

                override fun onSlide(bottomSheet: View, slideOffset: Float) {
                    binding.offsetText.text = getString(R.string.offset, slideOffset)
                }
            })
        }
    }

    override fun onBackPressed() {
        mBottomSheetBehavior?.let {
            if (it.state != BottomSheetBehavior.STATE_HIDDEN) {
                it.setState(BottomSheetBehavior.STATE_COLLAPSED)
            } else {
                super.onBackPressed()
            }
        }

    }

    companion object {
        fun getStateAsString(newState: Int): Int {
            when (newState) {
                BottomSheetBehavior.STATE_COLLAPSED -> return R.string.collapsed
                BottomSheetBehavior.STATE_DRAGGING -> return R.string.dragging
                BottomSheetBehavior.STATE_EXPANDED -> return R.string.expanded
                BottomSheetBehavior.STATE_HIDDEN -> return R.string.hidden
                BottomSheetBehavior.STATE_SETTLING -> return R.string.settling
            }
            return R.string.undefined
        }

        fun listApplications(context: Context): List<ApplicationInfo> {
            val flags = PackageManager.GET_META_DATA
            val installedApps: MutableList<ApplicationInfo> = ArrayList()
            val pm = context.packageManager
            val applications = pm.getInstalledApplications(flags)
            for (appInfo in applications) {
                if (appInfo.flags and ApplicationInfo.FLAG_SYSTEM != 1) {
                    installedApps.add(appInfo)
                }
            }
            return installedApps
        }
    }
}